### DISCUSSION TITLE
Sending client_id in token request when using private_key_jwt

### DISCUSSION THREAD

**Discussion by loikg** [keycloak/keycloak#19206] [Q&A] [answered]
Hi there,

I am working on integrating a OIDC Identity Provider in Keycloak using the OpenID Connect provider.
Unfortunately the idp I am using expect the `client_id` to be send as part of the request to `/token`. Something along the line of:

```
POST /token HTTP/1.1
Content-Type: application/x-www-form-urlencoded; charset=ISO-8859-1
Host: idp.com

client_assertion_type=xxx&code_verifier=xxx&client_id=xxx&code=xxx4&client_assertion=eyJraWQiO
```

According to the OAuth 2.0 spec https://datatracker.ietf.org/doc/html/rfc7521#section-4.2 `client_id` is optional:

> client_id
>       OPTIONAL.  The client identifier as described in [Section 2.2](https://datatracker.ietf.org/doc/html/rfc7521#section-2.2) of
>       OAuth 2.0 [[RFC6749](https://datatracker.ietf.org/doc/html/rfc6749)].  The "client_id" is unnecessary for client
>       assertion authentication because the client is identified by the
>       subject of the assertion.  If present, the value of the
>       "client_id" parameter MUST identify the same client as is
>       identified by the client assertion.

As far as i could find Keycloak does not provide a way to optionally send the `client_id`. Am I missing something or is there any way to configure Keycloak to send the `client_id`? 

Regards,

**Comment by justin-tay** [comment, marked as answer]
I also have this requirement. There is currently no option to send the `client_id` with the token request when using a client assertion. I have opened a pull request to add this feature.

- https://github.com/keycloak/keycloak/pull/21523

**Reply by loikg** [reply]
Awesome thanks. Great to know someone is working on this.
